File: frontend/src/types/learning.ts
export interface Curriculum {
  id: string;
  code: string;
  name: string;
  board_authority: string;
  education_level: string;
  country: string;
  description?: string;
  is_active: boolean;
}

export interface StudyMaterialDocument {
  id: string;
  user_id: string;
  title: string;
  source_type: string;
  file_path: string;
  file_size_bytes: number;
  status: 'queued' | 'processing' | 'completed' | 'failed' | 'retry';
  curriculum_id?: string;
  subject_id?: string;
  language: string;
  version: number;
  progress_percent: number;
  processing_stage: string;
  error_message?: string;
  page_count: number;
  content_hash?: string;
  metadata_json: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface Subject {
  id: string;
  curriculum_id?: string;
  name: string;
  slug: string;
  description: string;
  icon: string;
  category?: string;
  difficulty_level?: string;
  education_level?: string;
  academic_domain?: string;
  is_system?: boolean;
  is_enrolled?: boolean;
  order_index?: number;
  topic_count?: number;
  concept_count: number;
}

export interface StudentSubject {
  id: string;
  user_id: string;
  subject_id: string;
  enrollment_source: string;
  is_active: boolean;
  subject?: Subject;
}

export interface ProfileCompleteness {
  score: number;
  is_complete: boolean;
  missing_fields: string[];
}

export interface AcademicIdentity {
  education_level: string;
  education_category: string;
  grade_level: string;
  curriculum_id?: string | null;
  curriculum_name?: string | null;
  curriculum_code?: string | null;
  board_authority?: string | null;
  board_type?: string | null;
  stream?: string | null;
  program?: string | null;
  academic_domain: string;
  state_region?: string | null;
  institution?: string | null;
  degree?: string | null;
  department?: string | null;
  specialization?: string | null;
  academic_year?: string | null;
  preferred_language: string;
}

export interface MaterialsSummary {
  total_count: number;
  ready_count: number;
  processing_count: number;
  failed_count: number;
  recent_materials: StudyMaterialDocument[];
}

export interface LearningToolStatus {
  id: string;
  name: string;
  description: string;
  icon: string;
  route: string;
  status: 'active' | 'upcoming_prompt' | string;
  phase_label: string;
}

export interface SyllabusVersion {
  id: string;
  syllabus_id: string;
  version_number: number;
  document_id?: string | null;
  raw_extracted_json?: any;
  is_active: boolean;
  activated_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface Syllabus {
  id: string;
  user_id: string;
  title: string;
  academic_level: string;
  institution?: string | null;
  program_degree?: string | null;
  academic_year?: string | null;
  status: 'uploaded' | 'processing' | 'extracted' | 'confirmed' | 'archived' | string;
  created_at: string;
  updated_at: string;
  active_version?: SyllabusVersion | null;
  versions?: SyllabusVersion[];
}

export interface WorkspaceOverview {
  profile_completeness: ProfileCompleteness;
  academic_identity: AcademicIdentity;
  active_syllabus?: Syllabus | null;
  enrolled_subjects: Subject[];
  materials_summary: MaterialsSummary;
  learning_tools: LearningToolStatus[];
  starter_subjects_available: number;
  system_exploration_hint?: string;
}

export interface Lesson {
  id: string;
  module_id: string;
  title: string;
  slug: string;
  content_type: 'explanation' | 'example' | 'definition' | 'key_points' | 'visual' | 'exercise' | 'reading' | string;
  order_index: number;
  estimated_minutes: number;
  is_active: boolean;
}

export interface LessonDetail extends Lesson {
  content: string;
  module_title?: string;
  concept_name?: string;
  concept_slug?: string;
  subject_name?: string;
  subject_slug?: string;
  previous_lesson_slug?: string | null;
  next_lesson_slug?: string | null;
  personalized_context?: PersonalizedContext | null;
}

export interface LearningModule {
  id: string;
  concept_id: string;
  title: string;
  slug: string;
  description?: string;
  learning_objective?: string;
  difficulty_level: string;
  estimated_minutes: number;
  order_index: number;
  lesson_count: number;
  prerequisites: string[];
}

export interface LearningModuleDetail extends LearningModule {
  why_it_matters: string;
  simple_explanation: string;
  technical_explanation: string;
  visualization_type: string;
  experiment_type: string;
  simulation_config: Record<string, any>;
  application_notes: string;
  lessons: Lesson[];
  concept_name?: string;
  concept_slug?: string;
  subject_name?: string;
  subject_slug?: string;
}

export interface Concept {
  id: string;
  topic_id: string;
  name: string;
  slug: string;
  summary: string;
  short_description?: string;
  difficulty: string;
  difficulty_level?: string;
  order_index: number;
  module_count: number;
  has_simulation?: boolean;
  has_visualization?: boolean;
  has_practice?: boolean;
  has_lab?: boolean;
  has_mindmap?: boolean;
  learning_modes?: string[];
}

export interface ConceptDetail extends Concept {
  modules: LearningModule[];
  topic_name?: string;
  topic_slug?: string;
  subject_name?: string;
  subject_slug?: string;
}

export interface Topic {
  id: string;
  subject_id: string;
  name: string;
  slug: string;
  description?: string;
  order_index: number;
  concept_count: number;
  concepts?: Concept[];
}

export interface TopicDetail extends Topic {
  concepts: Concept[];
  subject_name?: string;
  subject_slug?: string;
}

export interface SubjectDetail extends Subject {
  topics: TopicDetail[];
}

export interface SimulationControl {
  id: string;
  label: string;
  min?: number;
  max?: number;
  step?: number;
  default: any;
  options?: string[];
}

export interface InteractiveSimulation {
  simulation_type: string;
  parameters: Record<string, any>;
  controls: SimulationControl[];
}

export interface PersonalizedContext {
  interest: string;
  domain: string;
  headline: string;
  analogy_explanation: string;
  real_world_application: string;
  related_domains: string[];
  relevance_score?: number;
}

export interface RecommendedTopic {
  concept: string;
  subject: string;
  slug: string;
  matched_interest: string;
  headline: string;
  summary: string;
  target_url: string;
}

export interface ConceptExploreResult {
  concept: string;
  domain: string;
  tagline: string;
  why_it_matters: string;
  simple_explanation: string;
  technical_explanation: string;
  simulation: InteractiveSimulation;
  practical_application: string;
  quick_check_question: string;
  quick_check_options: string[];
  quick_check_answer_index: number;
  personalized_context?: PersonalizedContext | null;
  available_perspectives?: PersonalizedContext[];
}

export interface BackendHealth {
  status: string;
  timestamp: string;
  project: string;
  tagline: string;
  system: {
    api: string;
    database: string;
    environment: string;
    version: string;
  };
  capabilities: Record<string, boolean>;
}

export interface StudentNote {
  id: string;
  user_id: string;
  academic_level: string;
  subject_id?: string | null;
  concept_id?: string | null;
  lesson_id?: string | null;
  title: string;
  content: string;
  tags: string[];
  source_reference?: string | null;
  is_pinned: boolean;
  is_archived: boolean;
  created_at: string;
  updated_at: string;
}

export interface PracticeQuestion {
  id: string;
  practice_set_id: string;
  concept_id?: string | null;
  lesson_id?: string | null;
  question_text: string;
  options: string[];
  points: number;
  difficulty: string;
  order_index: number;
}

export interface PracticeSet {
  id: string;
  concept_id?: string | null;
  lesson_id?: string | null;
  subject_id?: string | null;
  academic_level: string;
  title: string;
  description?: string | null;
  difficulty: string;
  questions_count: number;
  questions: PracticeQuestion[];
}

export interface QuestionSubmissionResult {
  question_id: string;
  question_text: string;
  selected_index: number;
  correct_index: number;
  is_correct: boolean;
  explanation: string;
  points_earned: number;
}

export interface PracticeResult {
  set_id: string;
  total_questions: number;
  correct_count: number;
  score_percentage: number;
  passed: boolean;
  results: QuestionSubmissionResult[];
}

export interface AcademicActivityLog {
  id: string;
  activity_type: string;
  academic_level: string;
  subject_id?: string | null;
  concept_id?: string | null;
  lesson_id?: string | null;
  title: string;
  description?: string | null;
  meta: Record<string, any>;
  created_at: string;
}

export interface SubjectProgress {
  subject_id: string;
  subject_name: string;
  progress_percent: number;
  completed_topics_count: number;
  total_topics_count: number;
}

export interface AcademicProgressOverview {
  academic_level: string;
  overall_progress_percent: number;
  completed_lessons_count: number;
  completed_quizzes_count: number;
  enrolled_subjects_count: number;
  study_minutes: number;
  subject_progress: SubjectProgress[];
}

File: backend/app/schemas/learning.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, ConfigDict


# ==============================================================================
# 1. Lesson Schemas (Level 5)
# ==============================================================================

class LessonBase(BaseModel):
    title: str
    slug: str
    content_type: str = "explanation"  # explanation, example, definition, key_points, visual, exercise, reading
    order_index: int = 0
    estimated_minutes: int = 5
    is_active: bool = True


class LessonRead(LessonBase):
    model_config = ConfigDict(from_attributes=True)
    id: str
    module_id: str


class LessonDetailRead(LessonRead):
    content: str
    module_title: Optional[str] = None
    concept_name: Optional[str] = None
    concept_slug: Optional[str] = None
    subject_name: Optional[str] = None
    subject_slug: Optional[str] = None
    previous_lesson_slug: Optional[str] = None
    next_lesson_slug: Optional[str] = None
    personalized_context: Optional[Dict[str, Any]] = None


# ==============================================================================
# 2. Learning Module Schemas (Level 4)
# ==============================================================================

class LearningModuleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    concept_id: str
    title: str
    slug: str = ""
    description: Optional[str] = None
    learning_objective: Optional[str] = None
    difficulty_level: str = "intermediate"
    estimated_minutes: int = 15
    order_index: int = 0
    is_active: bool = True
    why_it_matters: str = ""
    simple_explanation: str = ""
    technical_explanation: str = ""
    visualization_type: str = "interactive-canvas"
    experiment_type: str = "parameter-tuning"
    simulation_config: Dict[str, Any] = Field(default_factory=dict)
    application_notes: str = ""
    prerequisites: List[str] = Field(default_factory=list)
    lessons: List[LessonRead] = Field(default_factory=list)


class LearningModuleSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    concept_id: str
    title: str
    slug: str
    description: Optional[str] = None
    learning_objective: Optional[str] = None
    difficulty_level: str = "intermediate"
    estimated_minutes: int = 15
    order_index: int = 0
    lesson_count: int = 0
    prerequisites: List[str] = Field(default_factory=list)


class LearningModuleDetail(LearningModuleRead):
    lessons: List[LessonRead] = Field(default_factory=list)
    concept_name: Optional[str] = None
    concept_slug: Optional[str] = None
    subject_name: Optional[str] = None
    subject_slug: Optional[str] = None


# ==============================================================================
# 3. Concept Schemas (Level 3)
# ==============================================================================

class ConceptBase(BaseModel):
    name: str
    slug: str
    summary: str
    short_description: Optional[str] = None
    difficulty: str = "intermediate"
    difficulty_level: str = "intermediate"
    order_index: int = 0
    is_active: bool = True


class ConceptRead(ConceptBase):
    model_config = ConfigDict(from_attributes=True)
    id: str
    topic_id: str
    module_count: int = 0
    has_simulation: bool = False
    has_visualization: bool = False
    has_practice: bool = True
    has_lab: bool = False
    has_mindmap: bool = True
    learning_modes: List[str] = Field(default_factory=lambda: ["learn", "ask", "practice", "notes"])


class ConceptDetail(ConceptRead):
    modules: List[LearningModuleSummary] = Field(default_factory=list)
    topic_name: Optional[str] = None
    topic_slug: Optional[str] = None
    subject_name: Optional[str] = None
    subject_slug: Optional[str] = None


# ==============================================================================
# 4. Topic Schemas (Level 2)
# ==============================================================================

class TopicRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    subject_id: str
    name: str
    slug: str
    description: Optional[str] = None
    order_index: int = 0
    is_active: bool = True
    concepts: List[ConceptRead] = Field(default_factory=list)
    concept_count: int = 0


class TopicDetail(TopicRead):
    concepts: List[ConceptRead] = Field(default_factory=list)
    subject_name: Optional[str] = None
    subject_slug: Optional[str] = None


# ==============================================================================
# 5. Curriculum & Subject Schemas (Level 0 & 1)
# ==============================================================================

class CurriculumRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    code: str
    name: str
    board_authority: str
    education_level: str
    country: str = "India"
    description: Optional[str] = None
    is_active: bool = True


class SubjectRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    curriculum_id: Optional[str] = None
    name: str
    slug: str
    description: Optional[str] = None
    icon: str = "BookOpen"
    category: str = "Computer Science & Engineering"
    difficulty_level: str = "all-levels"
    education_level: str = "all-levels"
    academic_domain: str = "General"
    is_system: bool = True
    is_enrolled: bool = False
    order_index: int = 0
    is_active: bool = True
    topic_count: int = 0
    concept_count: int = 0
    topics: List[TopicRead] = Field(default_factory=list)


class SubjectDetail(SubjectRead):
    topics: List[TopicDetail] = Field(default_factory=list)
    curriculum: Optional[CurriculumRead] = None


class StudentSubjectRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    user_id: str
    subject_id: str
    enrollment_source: str = "student_selected"
    is_active: bool = True
    subject: Optional[SubjectRead] = None


class StudentSubjectCreate(BaseModel):
    subject_id: str
    enrollment_source: str = "student_selected"


# ==============================================================================
# 6. Experiential & Personalization Blueprints (Prompts 01-04 Compatibility)
# ==============================================================================

class ConceptExploreRequest(BaseModel):
    query: str = Field(..., min_length=2, max_length=200, description="Concept or question the student wants to understand")
    subject_hint: Optional[str] = Field(None, description="Optional subject domain hint")
    interest_hint: Optional[str] = Field(None, description="Optional student interest domain (e.g. Gaming, Cricket, Cars)")


class InteractiveSimulationSpec(BaseModel):
    simulation_type: str = Field(..., description="E.g., wave_compression, tree_traversal, kernel_filter")
    parameters: Dict[str, Any] = Field(default_factory=dict)
    controls: List[Dict[str, Any]] = Field(default_factory=list)


class ConceptExploreResponse(BaseModel):
    concept: str
    domain: str
    tagline: str
    why_it_matters: str
    simple_explanation: str
    technical_explanation: str
    simulation: InteractiveSimulationSpec
    practical_application: str
    quick_check_question: str
    quick_check_options: List[str]
    quick_check_answer_index: int
    personalized_context: Optional[Dict[str, Any]] = Field(None, description="Active tailored contextual perspective")
    available_perspectives: List[Dict[str, Any]] = Field(default_factory=list, description="All available interest perspectives for this concept")
